File: Modes/SpotifyMode.swift
/// Controller profile active when Spotify is the frontmost application.
///
/// All actions use the Spotify AppleScript dictionary — no keyboard shortcuts
/// needed. Playback control (play/pause, next/prev, seek, volume, shuffle,
/// repeat) is fully reliable via AppleScript on Mac.
///
/// - Note: "Save to library" is implemented as Cmd+S (Spotify's own keyboard
///   shortcut) rather than an AppleScript verb, because the `save` action is
///   not a first-class primitive in Spotify's scripting dictionary.
struct SpotifyMode: AppMode {

    let id           = "spotify"
    let displayName  = "Spotify"
    let sfSymbolName = "music.note"

    var buttonMap: [GamepadButton: Action] {
        [
            // ── Face buttons ──────────────────────────────────────────────

            .a: .appleScript(
                AppleScriptExecutor.Spotify.playpause,
                id: "spotify-play-pause",
                name: "Play / Pause"
            ),

            // Cmd+S = Spotify's native "Save to Library" shortcut.
            // More reliable than AppleScript for library actions.
            .b: .keystroke(
                .s,
                modifiers: .maskCommand,
                id: "spotify-save-track",
                name: "Save Track to Library"
            ),

            .x: .appleScript(
                AppleScriptExecutor.Spotify.toggleShuffle,
                id: "spotify-toggle-shuffle",
                name: "Toggle Shuffle"
            ),

            .y: .appleScript(
                AppleScriptExecutor.Spotify.toggleRepeat,
                id: "spotify-toggle-repeat",
                name: "Toggle Repeat"
            ),

            // ── D-Pad ─────────────────────────────────────────────────────

            .dpadUp: .appleScript(
                AppleScriptExecutor.Spotify.adjustVolume(by: Constants.Spotify.volumeStep),
                id: "spotify-volume-up",
                name: "Volume Up"
            ),

            .dpadDown: .appleScript(
                AppleScriptExecutor.Spotify.adjustVolume(by: -Constants.Spotify.volumeStep),
                id: "spotify-volume-down",
                name: "Volume Down"
            ),

            .dpadLeft: .appleScript(
                AppleScriptExecutor.Spotify.previousTrack,
                id: "spotify-previous-track",
                name: "Previous Track"
            ),

            .dpadRight: .appleScript(
                AppleScriptExecutor.Spotify.nextTrack,
                id: "spotify-next-track",
                name: "Next Track"
            ),

            // ── Shoulder buttons ──────────────────────────────────────────

            .leftShoulder: .appleScript(
                AppleScriptExecutor.Spotify.seekRelative(by: -Constants.Spotify.seekStepSeconds),
                id: "spotify-seek-back",
                name: "Seek Back 15s"
            ),

            .rightShoulder: .appleScript(
                AppleScriptExecutor.Spotify.seekRelative(by: Constants.Spotify.seekStepSeconds),
                id: "spotify-seek-forward",
                name: "Seek Forward 15s"
            ),

            // ── Menu buttons ──────────────────────────────────────────────

            // Start: bring Spotify to front (it's already frontmost in this mode,
            // but this is useful after mode-switching back from another app)
            .start: .openApp(
                bundleIdentifier: Constants.BundleID.spotify,
                id: "spotify-focus",
                name: "Focus Spotify"
            ),
        ]
    }
}

File: pocket_study_remote/modes/spotify_mode.py
from __future__ import annotations

from ..constants import BundleID, Spotify
from ..core.action import Action, apple_script, keystroke, open_app
from ..core.configurable_mode import ConfigurableMode
from ..executors.applescript_executor import AppleScriptExecutor


class SpotifyMode(ConfigurableMode):
    """
    Active when Spotify is the frontmost application.

    All playback actions use Spotify's AppleScript dictionary — fully
    reliable and background-safe.

    Buttons are configurable via Edit Keybindings.
    """

    id             = "spotify"
    display_name   = "Spotify"
    sf_symbol_name = "music.note"

    def _create_action_from_id(self, action_id: str) -> Action | None:
        """Create action from ID."""
        action_map = {
            "play_pause": apple_script(
                AppleScriptExecutor.Spotify.PLAYPAUSE,
                id="spotify-play-pause",
                name="Play / Pause",
            ),
            "next_track": apple_script(
                AppleScriptExecutor.Spotify.NEXT_TRACK,
                id="spotify-next-track",
                name="Next Track",
            ),
            "prev_track": apple_script(
                AppleScriptExecutor.Spotify.PREVIOUS_TRACK,
                id="spotify-previous-track",
                name="Previous Track",
            ),
            "volume_up": apple_script(
                AppleScriptExecutor.Spotify.adjust_volume(Spotify.VOLUME_STEP),
                id="spotify-volume-up",
                name="Volume Up",
            ),
            "volume_down": apple_script(
                AppleScriptExecutor.Spotify.adjust_volume(-Spotify.VOLUME_STEP),
                id="spotify-volume-down",
                name="Volume Down",
            ),
            "seek_forward": apple_script(
                AppleScriptExecutor.Spotify.seek_relative(Spotify.SEEK_STEP_SECONDS),
                id="spotify-seek-forward",
                name="Seek Forward",
            ),
            "seek_backward": apple_script(
                AppleScriptExecutor.Spotify.seek_relative(-Spotify.SEEK_STEP_SECONDS),
                id="spotify-seek-backward",
                name="Seek Backward",
            ),
            "shuffle": apple_script(
                AppleScriptExecutor.Spotify.TOGGLE_SHUFFLE,
                id="spotify-toggle-shuffle",
                name="Toggle Shuffle",
            ),
            "like": keystroke(
                "s", ["cmd"],
                id="spotify-save-track",
                name="Save Track to Library",
            ),
            "focus": open_app(
                BundleID.SPOTIFY,
                id="spotify-focus",
                name="Focus Spotify",
            ),
        }
        return action_map.get(action_id)
